### ISSUE TITLE
Add overscan (border) support

### ISSUE THREAD

**Issue by johnnovak** [dosbox-staging/dosbox-staging#1792]
# Overview

On real hardware, the actual drawing area (e.g. 320x200, 640x480, etc.) doesn't fill the whole screen but there's always a border around it (usually black). It is possible to set the colour of this "overscan area" to some other colour than black, and some games and applications use the border colour to convey information to the user. But even when the coloured border doesn't serve any "useful" purpose, it enhances the "retro feel" for people who still remember how real CGA/EGA/VGA output on a CRT looked like.

## CGA-like systems

On CGA and CGA-like systems (200-line EGA, Tandy, and PCjr), the overscan colour fills the whole screen, as you can see on the second and third image. There is no way to change the extents of the overscan border in software; the size is hardcoded.

![image](https://user-images.githubusercontent.com/698770/178980494-bc0ede02-a9c6-4618-a898-41a284138484.png)

![image](https://user-images.githubusercontent.com/698770/178983173-24ecdb0e-21fc-42d1-826b-41e7d222e59d.png)

![image](https://user-images.githubusercontent.com/698770/178978902-1384424c-092d-46aa-ab71-6670f6e1add2.png)

## VGA

On VGA, the overscan area is much smaller, usually just a few extra pixels, and most of the black border still remains. The actual size of the border can be controlled via software (there is a dedicated VGA register for it).

In the below video, note that every time something is selected in the menu, the screen goes full red, and then fades to black. When this happens, you can see a small stripe of red outside the visible menu area (the 320x200 actual content area); that's your border background colour on VGA.

https://youtu.be/LM1HYM_qSsw?t=84

# Examples

A non-exhaustive list of games that use the overscan area:

- Alley Cat
- Castle of Kroz
- Catacombs
- Caverns of Kroz 1, 2
- Commander Keen 1, 2, 3, 4
- Crystal Caves
- Dungeons of Kroz 1, 2
- Doom
- Duke Nukem 3D
- Final Crusade of Kroz
- Flashback
- Impossible Mission II (Tandy)
- Kingdom of Kroz 1, 2
- Lost Adventures of Kroz
- Return to Kroz
- Secret Agent
- Temple of Kroz
- Wolfenstein 3D

Applications that use the overscan area:

- FastTracker II

### Crystal Caves

> In the original DOS version, the game's border will change color from red to green when all crystals have been collected and the level can be exited. This feature is not emulated in the DOSBox emulator (which the official releases for Windows, Mac and Linux use), thus the player has to rely on only a sound cue to mark this.
([source](https://www.mobygames.com/game/crystal-caves))

### FastTracker II

When you toggle _Edit Mode_ by pressing Space, the border changes colour, and there's no other way to tell whether you are in _Edit Mode_ or not.

### DOOM, Wolfenstein 3D

When you get hit, not only the screen but also the border flashes red.

# Implementation

Currently, DOSBox Staging doesn't support overscan at all. DOSBox-X has limited support, and there was an old experimental [proof-of-concept hack](https://www.vogons.org/viewtopic.php?f=41&t=9814) that added basic support, but apparently it only ever worked in windowed mode.

Other emulators, such as FUSE, VICE and WinUAE/P-UAE, have been emulating overscan borders right from the start (but to be fair, overscan borders play a much bigger role on the ZX Spectrum, Amiga and C64 than on the PC).

There's some lengthy discussions on the Vogons forum about the exact extents of the overscan area on different graphics adapters and monitors, and ideas about possible implementations (see the references below). I don't think we need to be so super-detailed about it, and I don't think that 100% faithfulness to the original hardware matters either.

The user should simply be able to set the horizontal and vertical overscan borders in pixels or percentages relative to the actual drawing area, and that's it. The borders would always be drawn symmetrically around the actual output, and it would always be "additive", meaning that in windowed mode the window size would be increased by the overscan dimensions, and in fullscreen mode parts of the border would simply be clipped if they wouldn't fit into the screen. If the user wanted to see the full border in fullscreen mode, they would need to restrict the viewport size with `viewport_resolution` accordingly.

# References

https://www.vogons.org/viewtopic.php?p=328413#p328413
https://www.vogons.org/viewtopic.php?p=421895#p421895
https://www.vogons.org/viewtopic.php?f=32&t=38643&p=348920#p348920
https://www.vogons.org/viewtopic.php?f=41&t=49160&start=60

**Comment by sulix** [comment]
FYI: My [Commander Keen reimplementation project](https://github.com/sulix/omnispeak) supports this, albeit emulating only the VGA variant.

The [border size](https://github.com/sulix/omnispeak/blob/master/src/id_vl_private.h) we use for VGA 320×200 is:
- Left and right borders are 8px wide each.
- Top and bottom borders are 7 scanlines each, which, due to 320×200 being a line doubled mode, is 3½ px high each.

The latter can cause a few problems at 1× scale, but that can be worked around with some care.

**Comment by Jkapp76** [comment]
If it's a vote, I vote YES for this one! I really dig this old stuff. I like to keep the border smaller though... like with Commodore 64 Emulators too.

And, If you play Alley Cat, use the booter version or you won't get the awesome PCJr music.

I think Archon too had a color-changing border.

**Comment by johnnovak** [comment]
Yeah I'll add this at some point, but it's relatively low on my priority list. Ideally, I'd like to do it after @GranMinigun's pipeline refactoring stuff; easier than fighting the current mess.

> I like to keep the border smaller though... like with Commodore 64 Emulators too.

With my proposed solution you could set the exact desired border size — even just a few pixels if that's what you want. I see very little point in mimicking the monitor hardware exactly for this feature like one of the guys on Vogons tried in his super complex patch I liked above.

**Comment by johnnovak** [comment]
> And, If you play Alley Cat, use the booter version or you won't get the awesome PCJr music.

You seem to be a bit of an expert in all things PCjr, perhaps you'd like to help with this Wiki page about games that have the best video and/or sound on the PCjr or the Tandy:

https://github.com/dosbox-staging/dosbox-staging/wiki/Games-with-Tandy-&-PCjr-exclusive-graphics-and-sound

It's work in progress, so don't go crazy yet, but I'll ping you when I think it's ready. I'm gonna include a lot of other games from another source and reformat the whole thing.

**Comment by GranMinigun** [comment]
The easiest way to show overscan colour would be to set clear colour for both texture and GL outputs. Of course, that will paint the whole area around actual image, but I guess that's still better than nothing? Couple with limited viewport size to also get top and bottom borders. Specific border size is a bit more involved, but should be trivial enough still (`SDL_FillRect()`, and some fiddling with the core shaders for OpenGL, respectively).

Does DOSBox currently expose any overscan controls? That is, is it possible to get the overscan colour and know when it was set?

**Comment by johnnovak** [comment]
> The easiest way to show overscan colour would be to set clear colour for both texture and GL outputs. Of course, that will paint the whole area around actual image, but I guess that's still better than nothing? Couple with limited viewport size to also get top and bottom borders. Specific border size is a bit more involved, but should be trivial enough still (`SDL_FillRect()`, and some fiddling with the core shaders for OpenGL, respectively).

Yeah there are some solutions that seem easy but have some drawbacks. E.g. when the viewport resolution is set to 960x720 (that's typically what I use on my 1920x1080 screen), I really don't want the whole black area in fullscreen to change colour.

For example, when I use FastTracker II, in edit mode the border becomes red. You're frequently in edit mode, like 50% of the time when writing music. Having to stare at a 50% red screen is a no-go (FT II uses 640x480, so there would be lots of blank space in fullscreen with the typical 2x scaling).

Another consideration, I definitely want to border to be processed by the shaders as well (probably that will happen automatically even with setting the clear colour). There's two reasons for this:

- The obvious one, the border is part of the emulated screen, so for example if the shader implements screen geometry emulation, the border needs to be a part of that, naturally.
- For CRT shaders that emulate bloom and glow (so all good ones), the glow effects are currently cut off at the edges of the emulated screen. So ideally we'd have a say 10px padding around the image, e.g. for 320x200 we'd be rendering into a 340x240 texture, with the image centered to the texture. This will become an even bigger problem once we have proper multi-pass shaders, as the best RetroArch CRT shaders have much better blow, gloom and halation emulation that extend much further around the pixels (these are impossible to do without multipass, at least not efficiently; the performance costs to do this in a single-pass are astronomical).

If I were to implement this now, I'd do it like this probably:

- `border_size` – Width of the border expressed as _percentage_ of the _width_ (!) of the DOS framebuffer. Default would be `2.5%`, which gives you 8px border around a 320x200 screen, 16px around a 640x480 screen, etc. (You need percentages/proportions instead of absolute pixels so it works consistently across all display modes.)

- `border_mode` — two options:

    -  `inner` — The border is included in the viewport size, making the drawable image area smaller. For example, for a 320x200 VGA game and the above default border setting, the actual drawable area would become 304x184, and the border would _always_ be fully visible inside the viewport since it's included in it.
    - 
    - `outer` — This is the more interesting setting and what we should default to. The border is applied on top of the specified viewport resolution, with the extra twist that the border is "allowed to be clipped" if it doesn't fit within the extents of the window or screen. The VICE C64 emulator implements this as one of its border mode options. In principle, we render into a texture "expanded" by the border dimensions (so for 320x200 we render into 336x216, with the border colour set, and the actual drawable image centered), _but_ then we center the image as if were just 320x200, letting the border to be clipped! It's easier to illustrate this via a few images.
  
  
`viewport_resolution` was set to `640x480` for these example image. Note the extents of the drawable area is unchanged in all example, it's always 640x480, and the border is always applied around it as extra padding, and clipped if it doesn't fit into the window.

### windowresolution == viewport_resolution — border fully clipped

<img width="752" alt="image" src="https://user-images.githubusercontent.com/698770/215233618-144b40b7-5f86-4c5a-a875-1e863b5dd31a.png">


### windowresolution > viewport_resolution — border not clipped

Same scenario in fullscreen mode.

![border1](https://user-images.githubusercontent.com/698770/215232554-522019a9-f666-4133-925a-96ace70798f9.png)

### windowresolution > viewport_resolution (vertically only) — border is not clipped vertically, but is clipped horizontally 

![border2](https://user-images.githubusercontent.com/698770/215232555-fcb2532a-9bd2-426e-affb-11970ab0dee3.png)

### variation of the above — the border is not clipped horizontally, but is clipped a little bit vertically because it doesn't fit into the window

![border3](https://user-images.githubusercontent.com/698770/215232558-0f6a876e-0928-4837-b179-789fce952e60.png)

Anyway, I think the images and my above description should make it clear how I envisage this to work. In the end, I think it will be easiest if I implement it 😄 I'm quite convinced this is the best way to implement border support 😄 

> Does DOSBox currently expose any overscan controls? That is, is it possible to get the overscan colour and know when it was set?

You can read it from certain CGA/EGA/VGA registers — it is easy to get that info.

<img width="1168" alt="image" src="https://user-images.githubusercontent.com/698770/215232541-62394132-2733-495f-a2c5-4669ba68c26a.png">

**Comment by GranMinigun** [comment]
> probably that will happen automatically even with setting the clear colour

Not the case, because clearing backbuffer doesn't equal drawing to it with a shader.

> You can read it from certain CGA/EGA/VGA registers — it is easy to get that info.

Great. Then it's only a matter of improving rendering pipeline.

**Comment by johnnovak** [comment]
> Not the case, because clearing backbuffer doesn't equal drawing to it with a shader.

Yeah, so everything what I wrote in my long post above stands then.

**Comment by sulix** [comment]
Personally, I suspect the _most accurate_ way to handle this is to have the VGA code just generate a larger image with the border included. The exact size of the border could easily depend on which video card is being emulated. (Though, as mentioned, you'd need to have the VGA code do line-doubling for some modes, which could cause problems for some setups. You could just disable line doubling and round the border size if scaling is disabled / the window is too small, though.)

The way I handle it in my Commander Keen project is by using the clear colour, but rendering to a texture (which then undergoes the final scaling). But I was rendering-to-texture as part of my two-stage scaling process anyway, so it has no extra cost: I suspect that DOSBox doesn't, so this might be slower.

Regardless, it's probably worth looking at @NY00123's [old patches on Vogons](https://www.vogons.org/viewtopic.php?p=327934#p327934). As mentioned, I think having the overscan border be a part of the VGA code is probably the way to go about it, so I'd probably start by looking [here](https://www.vogons.org/viewtopic.php?p=329798#p329798).

Of course, that wouldn't get you the border clipping you describe without some significant extra work…

**Comment by johnnovak** [comment]
> Personally, I suspect the _most accurate_ way to handle this is to have the VGA code just generate a larger image with the border included. The exact size of the border could easily depend on which video card is being emulated. (Though, as mentioned, you'd need to have the VGA code do line-doubling for some modes, which could cause problems for some setups. You could just disable line doubling and round the border size if scaling is disabled / the window is too small, though.)
> 
> The way I handle it in my Commander Keen project is by using the clear colour, but rendering to a texture (which then undergoes the final scaling). But I was rendering-to-texture as part of my two-stage scaling process anyway, so it has no extra cost: I suspect that DOSBox doesn't, so this might be slower.
> 
> Regardless, it's probably worth looking at @NY00123's [old patches on Vogons](https://www.vogons.org/viewtopic.php?p=327934#p327934). As mentioned, I think having the overscan border be a part of the VGA code is probably the way to go about it, so I'd probably start by looking [here](https://www.vogons.org/viewtopic.php?p=329798#p329798).
> 
> Of course, that wouldn't get you the border clipping you describe without some significant extra work…

Yeah so that is what I proposed, to render the border to texture, minus using the "exact" approach from that old patch. That unnecessarily complicates things for no good reason.

The clipping thing isn't difficult; it's more difficult to explain it than to implement. Just a few extra lines of code to the logic that places the texture within the viewport, plus all the windowing logic should use the borderless dimension. Easy to do in practice.

**Comment by Jkapp76** [comment]
I use the Hoxs Commodore 64 emulator and it allows you to set the border to just the right and left sides. This effectively makes your widescreen monitor the correct oldschool dimensions and gives you the colored border at the left and right sides.
This is my favorite way to keep it for C64 games.

**Comment by johnnovak** [comment]
> I use the Hoxs Commodore 64 emulator and it allows you to set the border to just the right and left sides. This effectively makes your widescreen monitor the correct oldschool dimensions and gives you the colored border at the left and right sides. This is my favorite way to keep it for C64 games.

With the `outside` border option you'd get exactly that in fullscreen.

**Comment by johnnovak** [comment]
> I recommend ripping the band-aid off and doing this the right way.

Well, I don't agree with any of that. What _practical benefits_ would this extremely hardware-accurate approach bring? Some games/apps use the border to convey information, that's all. It's nice to be able to see that, because we currently don't have that visual feedback. We don't want to make this more complex than that—not even DOSBox-X does that...

Implementing a super-accurate-down-to-the-cycle level VGA emulation is not the goal of DOSBox _by far_. This would add a lot of unnecessary complexity for the code for no good reason, and it would be not in line with our goals. Please see our goals here: https://dosbox-staging.github.io/about/

Screwing around with the ideal 4:3 aspect ratio because of the width of the border... Man, no one cares for that stuff. While _technically_ the border size might affect the actualy image size/aspect ratio a _little bit_, it's a moot point. Tell me _one_ example where the app/game uses some extreme border settings which would _substantially_ alter the perceived aspect ratio, and that was actually the intent of the developer. I would bet good money on it that the number of such programs is exactly zero...

**Comment by kcgen** [comment]
> @anikom15,
> I recommend ripping the band-aid off and doing this the right way.

The VGA draw code is compatible with over 10,000 DOS games spanning 25+ years of hardware evolution - and is more compatible than any single piece of original PC hardware in existence (https://gona.mactar.hu/DOS_TESTS/); so it's pretty hard for me to sit back in an armchair and claim that *"DOSBox is doing it wrong"*.

With SDL now providing a hardware-accelerate "flexible" canvas, there is no doubt logic and software-scaling methods that can be eliminated from  vga_draw and render.cpp, and perhaps even re-introduce some border logic. 

Is that something you're interested in digging into? Touching this code  while maintaining performance.. without regressions.. without (further) complexifying the code.. and running countless regression tests to ensure correct behavior is a tall order! 

My own effort in combining `vgadraw` with `svga_*` is itself a partial failure.. I've introduced regressions and the code isn't any better off! (and I'm feeling like Brundle Fly after all this).  

There's a lot of room for improvement, but there's a huge amount of risk for failure and regression.  I appreciate the confidence, and yes - like @johnnovak said, if you can refactor it and make it all run: then my hats off to you! Join the effort and dig in :spaghetti:

**Comment by johnnovak** [comment]
> There's a lot of room for improvement, but there's a huge amount of risk for failure and regression. I appreciate the confidence, and yes - like @johnnovak said, if you can refactor it and make it all run: then my hats off to you! Join the effort and dig in 🍝

Yeah, please clean up and simplify the graphics card emulation code @anikom15, plus add the border stuff in a clean way, while not breaking any of the 10k+ supported games. If you can do it, we'll merge it in, of course!

But screwing around with the 1:1 PAR in modes that are _supposed_ be 1:1 PAR is a bit crazy... I personally won't have any of that.

**Comment by johnnovak** [comment]
I'm honestly curious (no irony) where would the "correct" border implementation have any tangible benefits over the proposed simplified approach where just pick a border width you like, basically? Can you list those programs? In my experience, very few programs use the border colour to convey some information, and for those that do, it does not matter one bit whether the border is 5px, 10px, 20px, or whatever else. Overall, probably 1% of all programs use the border in any meaningful way, making this a rather niche and quite low priority feature anyway.

**Comment by johnnovak** [comment]
Note hardware accuracy and "emulation as documenting hardware accurately" is not among our goals, by far. It was never among the goals of the original DOSBox either. Have you ever seen the VGA code? 😎

I'm all about preservation too, but I think you're barking up the wrong tree here. In DOSBox Staging it's always a balance between accuracy **_and_** tangible benefits to the end user, which is 99% of the time people playing old DOS games. Not developing new software, not trying to make the hardware do never-before-seen-things (that would require an 1:1 FPGA re-creation of the hardware anyway), not even running demos that do low-level trickery correctly. Main focus is playing games, that's about it. Getting most apps to work correctly is a bonus we get "for free", kinda.

Accurate sound emulation that affects all games and audio being a very important, crucial part of the _gaming_ experience—yes, by all means! Accurate border emulation where the "accurate" border is not an essential part of the experience, and less than 1% of games use the border meaningfully for anything, while complicating the already complicated-beyond-imagination graphics card code (note it's not "accurate" emulation of CGA/EGA/VGA/etc.—it's a mismash of all things that does all gfx card emulation in "one spot"... very messy, and by no means an "accurate representation of hardware". It's "accurate" enough to run 99% of all games correctly, and that's it, it was never any different in DOSBox land).

I'm asking again, do you have a list of games where the accurate border emulation makes a big difference for the average user who's not super into accurate border? 😄 Please share the list of games and appications. I've listed some apps at the top where the border conveys info—the "simplified" border emulation I proposed would take care of those. As far as I see it, that approach is sufficient and hits the right balance between complexity, effort, and benefit.

Specifically, from our about page, please familiarise yourself with the goals of the project before raising similar tickets and comments (and read the _full thing_; the non-goals section refers to the "above goals" in the "goals" section): https://dosbox-staging.github.io/about/

![image](https://github.com/dosbox-staging/dosbox-staging/assets/698770/b4db0918-bc97-44f8-a566-6007b1572fe5)

**Comment by johnnovak** [comment]
Having said all that, dig in, and show us how it's done in an easy, accurate, and maintainable way
 It has to work correctly in windowed and fullscreen, with shaders, in the texture output mode, handling borders in the screenshots need to be worked out, etc. Looking forward to your design and PR, and if it's great, we'll sure as hell merge it!

Otherwise, I'll do as planned.

**Comment by johnnovak** [comment]
> cycle accuracy is in itself a myth, but I digress

Yes and no. It makes a lot of sense and it's practically mandatory to accurately emulate the timings of the various custom chips in many more advanced home computers, e.g. the Commodore Amiga and the Commodore 64. Without that most games and demos would simply not run. Same for cycle-accurate, flux level floppy emulation; it's mandatory for emulating many copy protection mechanisms.

It's more like a sliding scale, and the PC architecture is a lot less advanced and much more decoupled, making a much looser and less accurate approach feasible. But you're right, I always smirk when people say "cycle-accurate" in conjunction with PCem or 86box. They're far from it in all aspects of the emulation, but of course they emulate a whole PC more accurately than the really loose and abstract DOSBox approach with "adaptive, selective" accuracy of specific areas (OPL -- cycle accurate; MT-32 emu -- 95%+ as hardware; VGA -- quite loose but most stuff works; CPU -- emulating a very abstract generic x86 CPU with a very simple MIPS approximation, etc.)

Then there are the Macintosh emulators which are really simple as there was barely any hardware in those early machines :trollface:

**Comment by johnnovak** [comment]
> Unfortunately, I have seen many resources, once easily found across the internet, disappear quietly during the past several years. 

By the way @anikom15 if you care about preserving such info, register an archive.org account and start uploading mirrors/dumps of existing sites that contain useful info about retro computing. I'm doing the same, just I'm limited by available time...

**Comment by dbalsom** [comment]
I would suggest not thinking about the border as something separate - rather think of a display field in totality, and the portion we display being some "aperture" into that field.  So for CGA, for example, the entire display field is 912x262 (* ~60Hz = ~14Mhz)

We can see as much or as little border as we please by defining an aperture into this field, in this case, I chose a value of 704x224 as the "accurate" setting, just based on aesthetics and a few photographs of scenes IBM 5153 monitors. It need not be perfect, every monitor was adjusted differently and the picture was not always centered well.

![cga_aperture](https://github.com/dosbox-staging/dosbox-staging/assets/7229541/4b0cb388-349b-4226-87b6-aa68d2f25f43)

The essential bit is not trying to choose a resolution from CRTC registers. The resolution is derived from the dot clock, which is fixed on MDA/CGA, one of two values on EGA/VGA, potentially more on SuperVGA, etc. But the basic method is adaptable to any raster device, although at times your "screen-accurate" aperture may be larger than the field in which case you would pad with black.

**Comment by FeralChild64** [comment]
It will have to be considered as something at least partially separate, or else the seamless mouse integration won’t work properly - there will be a strange effect when moving mouse in/out of the screen drawing area.

**Comment by johnnovak** [comment]
> I would suggest not thinking about the border as something separate - rather think of a display field in totality, and the portion we display being some "aperture" into that field. So for CGA, for example, the entire display field is 912x262 (* ~60Hz = ~14Mhz)

Thanks for stopping by @dbalsom. I'm aware of your interests in accurate emulation, but I need to consider practicality here and how this would work _together_ (not against) our existing features. At the end of the day, sure we could implement anything, accurate borders to the max etc... but to what end? Some programs convey important info in the border, and that's it. I'm not interested in doing this 100% accurately, just doing the bare minimum that conveys this info in good old DOSBox tradition 😎 Not interested in those 5 games either that use border tricks to shake the screen, etc. People can live without that, and the code will be much simpler 😄 

I definitely do want to implement this as something "tacked-on" at the end of the frame render time. Literally like this in pseudo-code:

1. Render frame normally
2. Check current border colour
3. Extend frame by configured border amount, fill it with colour.
4. Present extended frame. In some border modes, the border can be partially or fully outside of the screen.

We most definitely won't suddenly treat say 640x480 as 645x485 (just making things up and assuming a 5px border). Like I said, I don't want to complicate the code and punish everybody with complexity and weird display resolutions just because 10-20 games use the border for something interesting...

**Comment by kcgen** [comment]
Thanks for the very understandable explanation, @dbalsom; and we couldn't ask for more battle-hardened insights!

@johnnovak, my guess is that treating the display as a 2D field that's scanned over at the given MHz rate would mean that line timings would become a lot more accurate, and perhaps some simplicity would fall out of this approach to. 

No doubt this change would ripple outward (having to handle or expect a large outbound frame at the rendering, shader, mouse-edges, etc), but my guess this would have to be parameterized (perhaps more so than it currently is). 

Just thinking about possibilities.

**Comment by johnnovak** [comment]
> @johnnovak, my guess is that treating the display as a 2D field that's scanned over at the given MHz rate would mean that line timings would become a lot more accurate, and perhaps some simplicity would fall out of this approach, too.

The current code must do that, otherwise thing would misbehave, I'm pretty sure. It's just only the visible part is rendered.

**Comment by dbalsom** [comment]
> @johnnovak, my guess is that treating the display as a 2D field that's scanned over at the given MHz rate would mean that line timings would become a lot more accurate, and perhaps some simplicity would fall out of this approach, too.

There is a surprising elegance to it. Whether you render as an indexed intermediate buffer or directly as RGBA, when you tick a character clock you are incrementing your buffer pointer by as many bytes as the character width (8 for CGA/EGA, 9 for MDA/VGA) or 4x for RGBA. When you reach hsync you reset your pointer position to the next scanline by the field stride, and vsync is just resetting your pointer the start of the buffer again. 

Nobody needs to be punished by obscure video sizes unless they desire to see them; the 'cropped' aperture values provide the traditional, regular resolutions we are all familiar with, 640x200, 320x200, 640x480, and so on, and if nothing else, they would be safe fallbacks for features that require them.  But I am not attempting to tell you how you should design things or what does or does not meet your goals - just suggesting an approach that worked for me and I think has some positive benefits besides emulation of esoterica.

**Comment by johnnovak** [comment]
> > @johnnovak, my guess is that treating the display as a 2D field that's scanned over at the given MHz rate would mean that line timings would become a lot more accurate, and perhaps some simplicity would fall out of this approach, too.
> 
> There is a surprising elegance to it. Whether you render as an indexed intermediate buffer or directly as RGBA, when you tick a character clock you are incrementing your buffer pointer by as many bytes as the character width (8 for CGA/EGA, 9 for MDA/VGA) or 4x for RGBA. When you reach hsync you reset your pointer position to the next scanline by the field stride, and vsync is just resetting your pointer the start of the buffer again.
> 
> Nobody needs to be punished by obscure video sizes unless they desire to see them; the 'cropped' aperture values provide the traditional, regular resolutions we are all familiar with, 640x200, 320x200, 640x480, and so on, and if nothing else, they would be safe fallbacks for features that require them. But I am not attempting to tell you how you should design things or what does or does not meet your goals - just suggesting an approach that worked for me and I think has some positive benefits besides emulation of esoterica.

Thanks for the clarification @dbalsom, I think I slightly misunderstood your previous post.

Yes, I think this is probably the best approach from the rendering point of view. The rest of the concerns I listed are unrelated to the pure rendering path, so my bad conflating two related but actually quite separate things 😅 

I original idea (crappy idea) was to copy the rendered image into a larger image to add the buffer. The next idea was to "oversize" the render buffer by the configured border amount, then either pre-fill that with the border colour, or insert some special logic to fill in the border colour at the start and end of lines. Well, basically what you're suggesting is a variation of that approach, so yeah, we're in agreement.

**Comment by dbalsom** [comment]
> It will have to be considered as something at least partially separate, or else the seamless mouse integration won’t work properly - there will be a strange effect when moving mouse in/out of the screen drawing area.

I'm not entirely familiar with this feature (I'm really curious how you sync up mouse deltas between guest and host...) but seems like maintaining a transformation matrix or such would not be infeasible - you could have a "display aperture" and an "input aperture" and a simple function to convert "display aperture" to "input aperture" coordinates.

**Comment by johnnovak** [comment]
> I'm really curious how you sync up mouse deltas between guest and host...)

Well, we can't, in the general sense. Games/programs are free to do whatever, so most of the time we have no idea where the mouse pointer is. The seamless mode really only works reliably with the VMware protocol in Windows 3.x that we support as the driver can report absolute positions in that specific protocol. But you need 3rd VMware Win 3x. mouse drivers for that.

**Comment by dbalsom** [comment]
![image](https://github.com/dosbox-staging/dosbox-staging/assets/7229541/67061f90-c1f3-4889-9ee8-512635770fd8)

here's a cleaned-up diagram I made to help illustrate the 'field buffer'. This is a standard NTSC diagram to scale, just doubled in height so it doesn't look squished.  The idea is that we reset the pointer to the next scanline at the end of the blue bar, and we reset the pointer to the start of the buffer at the end of the purple one, and so on. 

This moves the left overscan to the left side of the buffer, and the top overscan to the top, respectively.  Then we just choose our aperture (sub-rectangle) into the field.  These apertures are fixed - as the few games that actually bother to do so adjust the CRTC registers, stuff will move around in the field, but our aperture remains fixed, and the adjustments and screen-shake effects "just work"
